### ISSUE TITLE
Manage ISC BIND DNS and DHCP

### ISSUE THREAD

**Issue by lindhor** [cockpit-project/cockpit#8456]
It would be great to be able to use cockpit to manage BIND DNS and DHCP via Cockpit if it has been installed. Do you think it would be reasonable to have such functionality in there or would it be preferred to have such tools outside Cockpit?

For DNS it could be used for things like managing resource records in zone files, add/remove zones, manage the named service, get stats, initiate zone transfers, generate signatures, manage settings in named.conf, manage sortlists etc.

For DHCP it could be used to manage static and dynamic leases, handle scopes, get stats, warn before scopes get full, manage DHCP options, manage the dhcpd service etc.

**Comment by lindhor** [comment]
Great! We are discussing investing time in it at the moment. It will be quite a lot of work, so we need to wait for a decision. Hope that we may have an answer the upcoming weeks.

**Comment by stefwalter** [comment]
Once you know, poke @garrett or @andreasn here, and they can help with design and a basic idea of where this would live in Cockpit.

**Comment by garrett** [comment]
I like your idea!

To me, this feels like it is an excellent candidate to be implemented as a server application.

Reasons:

1. While it initially feels a bit top-level, as it affects several hosts, it really is editing one specific host that others connect to (instead of managing many machines at once). Therefore, you're editing one host and it should go in host navigation (sub-navigation area for hosts).
2. Not all machines need this configuration. This is what's driving me to suggest an implementation as a server application.
3. Server applications get a special spot in the host navigation.

Server apps were released last October, in Cockpit 152. A few details are on the 152 release blog post @ http://cockpit-project.org/blog/cockpit-152.html

I'm not sure where the developer documentation is; @mvollmer should know.

Meanwhile, there are a few example apps:
- [cockpit-sosreport](https://github.com/cockpit-project/cockpit/tree/master/pkg/sosreport) (in-tree app)
- [cockpit-freeipa](https://github.com/mvollmer/cockpit-app-freeipa) (standalone app)
- [cockpit-app-samba-ad](https://github.com/abbra/cockpit-app-samba-ad) (third-party standalone app, based on cockpit-freeipa)

---

If not implemented as an application, then (a slightly tweaked version of) the great advice Stef gave in PR #5980 is a good guide on where it should go:

> - System info
> - Logs for troubleshooting
> - Configuring major subsystems: Storage, Networking
> - Stuff running on the system: Containers and Virtual Machines
> - Implementation details: Admin accounts, Services/Units
>   - These tools are sorted alphabetically

Basically, the "Stuff running on the system" group is where this would live, if not an app.

_Note: At initial glance, it might make sense to group into networking (similar to [what we're doing with upcoming firewall configuration](https://github.com/cockpit-project/cockpit/wiki/Feature:-Firewall)), but this doesn't concern the machine's own networking config, it (as you know) lets you control a specific service for others to use — so it shouldn't be put under networking._

**Comment by lindhor** [comment]
Thanks for all info. I will let the team know. By the way, do any of you have a very rough guesstimate how much time it would take an experienced developer like yourself to make an app such like this that is fairly complex ie including everything necessary.

**Comment by martinpitt** [comment]
We usually don't go about a large problem like this by defining, designing, and implementing "everything necessary" in one go, as that would take way too much time. Particularly with a new page/project like this it's much better to define a absolutely minimum things that make the page fulfill one basic use case - even just displaying the current state. Then iterate on the design, write tests, etc. and then add features step by step. This is how we develop cockpit itself.

Building a reasonably rich UI for this will take several months if you go through the full procedure of doing user testing, getting feedback, iterating on designs and use cases, tests, etc. The actual implementation wrt. implementing the UI and writing updated config files is probably not too hard . The main difficulty is parsing the existing configuration: I'm not terribly familiar with bind, but from what I remember it has a very rich configuration file structure and no other API (D-Bus, REST, etc). So the main difficulty for this is that your code needs to be able to understand this to avoid breaking an existing configuration. Or at least understand it enough to know when it can't manage the file because it was customized "too much".

**Comment by davewthompson** [comment]
Random comment, but I wonder if https://github.com/adamcooke/bound could be helpful for you in any way?

**Comment by lindhor** [comment]
Sorry to say, but we will not be able to go ahead with this due to lack of resources. Still think it might be a good idea though if somebody else is up for it.

**Comment by martinpitt** [comment]
I agree this should be implemented as an external Application (see [above comment](https://github.com/cockpit-project/cockpit/issues/8456#issuecomment-360536737)).

**Comment by fxrsliberty** [comment]
This needs to be looked at again... This and a couple of other standard system services , DHCP , NTP, TFTP, etc..

**Comment by alvinc-git** [comment]
I think maybe this should be resurrected.  There is also the opportunity of support ISC KEA and helping with the transition from ISC DHCPd.

Especially since I prefer cockpit to webmin (which has some support for this already).

**Comment by fxrsliberty** [comment]
+1 on this, it's should be a little easier to build a DHCP server for
isolated vlan specific tasks.

**Comment by Macr0Nerd** [comment]
+1 It would be extremely helpful as Cockpit is already my one stop shop for managing my servers. Being able to configure Bind from there would be extremely helpful.

**Comment by martinpitt** [comment]
Yes, nobody said that it wasn't a good idea. The comment above was merely expressing that we don't have time/priorities for this in the cockpit team itself, but we are happy to guide someone else with writing it.

**Comment by Macr0Nerd** [comment]
@martinpitt Do you mind elaborating on what would be necessary to write it? I am as far from a web developer as you can get but I am a fairly competent devops engineer and systems/network admin. I might be able to help at least get the ball rolling, especially on backend operation.

**Comment by martinpitt** [comment]
The entry point to creating a new cockpit page is here: https://cockpit-project.org/external/wiki/Contributing.html#further-reading , especially the first two posts.

**Comment by Gideon2361** [comment]
While not directly related to Cockpit itself, there is a module within Webmin that already does this...perhaps the Webmin module could be used as a springboard for a Cockpit implementation?  At the very least, perhaps those coming here looking for a Cockpit version (as I just did) might benefit from knowing Webmin does offer BIND support.

**Comment by Gideon2361** [comment]
A direct link to the section of Webmin source related to BIND management in case anyone with the interest in writing a Cockpit module wishes to see how Webmin managed it.

https://github.com/webmin/webmin/tree/master/bind8
